### DISCUSSION TITLE
DeFi StableCoin: DecentralizedStableCoin.sol

### DISCUSSION THREAD

**Discussion by hard1k7724** [Cyfrin/foundry-full-course-cu#1714] [Q&A] [answered]
Getting this error in DecentralizedStableCoin.sol the ERC20 one : No arguments passed to the base constructor. Specify the arguments or mark "DecentralizedStableCoin" as abstract.solidity(3415)

Here is the code:
//SPDX-License-Identifier:MIT

// This is considered an Exogenous, Decentralized, Anchored (pegged), Crypto Collateralized low volitility coin

// Layout of Contract:
// version
// imports
// interfaces, libraries, contracts
// errors
// Type declarations
// State variables
// Events
// Modifiers
// Functions

// Layout of Functions:
// constructor
// receive function (if exists)
// fallback function (if exists)
// external
// public
// internal
// private
// view & pure functions

pragma solidity ^0.8.18;

import {ERC20Burnable, ERC20} from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";

/*
 * @title DecentralizedStableCoin
 * @author Patrick Collins
 * Collateral: Exogenous (ETH & BTC)
 * Minting: Algorithmic
 * Relative Stability: Pgged to USD
 *
 * This is the contract meant to be governed by the DSCEngine. This contract is just the ERC20 implementation of our stablecoin system.
 *
 */
contract DecentralizedStableCoin is ERC20Burnable, Ownable {
    error DecentralizedStableCoin__MustBeMoreThanZero();
    error DecentralizedStableCoin__BurnAmountExceedsBalance();
    error DecentralizedStableCoin__NotZeroAddress();

    constructor() ERC20("DecentralizedStableCoin", "DSC") {}

    function burn(uint256 _amount) public override onlyOwner {
        uint256 balance = balanceOf(msg.sender);
        if (_amount <= 0) {
            revert DecentralizedStableCoin__MustBeMoreThanZero();
        }
        if (balance < _amount) {
            revert DecentralizedStableCoin__BurnAmountExceedsBalance();
        }
        super.burn(_amount);
    }

    function mint(
        address _to,
        uint256 _amount
    ) external onlyOwner returns (bool) {
        if (_to == address(0)) {
            revert DecentralizedStableCoin__NotZeroAddress();
        }
        if (_amount <= 0) {
            revert DecentralizedStableCoin__MustBeMoreThanZero();
        }
        _mint(_to, _amount);
        return true;
    }
}

**Comment by EngrPips** [comment]
```js
//SPDX-License-Identifier:MIT

// This is considered an Exogenous, Decentralized, Anchored (pegged), Crypto Collateralized low volitility coin

// Layout of Contract:
// version
// imports
// interfaces, libraries, contracts
// errors
// Type declarations
// State variables
// Events
// Modifiers
// Functions

// Layout of Functions:
// constructor
// receive function (if exists)
// fallback function (if exists)
// external
// public
// internal
// private
// view & pure functions

pragma solidity ^0.8.18;

import {ERC20Burnable, ERC20} from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";

/*

@title DecentralizedStableCoin
https://github.com/author Patrick Collins
Collateral: Exogenous (ETH & BTC)
Minting: Algorithmic
Relative Stability: Pgged to USD
This is the contract meant to be governed by the DSCEngine. This contract is just the ERC20 implementation of our stablecoin system.
*/
contract DecentralizedStableCoin is ERC20Burnable, Ownable {
error DecentralizedStableCoin__MustBeMoreThanZero();
error DecentralizedStableCoin__BurnAmountExceedsBalance();
error DecentralizedStableCoin__NotZeroAddress();

constructor() ERC20("DecentralizedStableCoin", "DSC") {}

function burn(uint256 _amount) public override onlyOwner {
    uint256 balance = balanceOf(msg.sender);
    if (_amount <= 0) {
        revert DecentralizedStableCoin__MustBeMoreThanZero();
    }
    if (balance < _amount) {
        revert DecentralizedStableCoin__BurnAmountExceedsBalance();
    }
    super.burn(_amount);
}

function mint(
    address _to,
    uint256 _amount
) external onlyOwner returns (bool) {
    if (_to == address(0)) {
        revert DecentralizedStableCoin__NotZeroAddress();
    }
    if (_amount <= 0) {
        revert DecentralizedStableCoin__MustBeMoreThanZero();
    }
    _mint(_to, _amount);
    return true;
}
}
```

**Reply by EngrPips** [reply]
Hello @Juice-07, It seems as though the `Ownable` contract explicitly needs a constructor argument. Can you please open the `Ownable` contract and confirm that?

**Reply by hard1k7724** [reply]
@EngrPips , Sure here is the constructor of the contract :

 abstract contract Ownable is Context {
    address private _owner;

    /**
     * @dev The caller account is not authorized to perform an operation.
     */
    error OwnableUnauthorizedAccount(address account);

    /**
     * @dev The owner is not a valid owner account. (eg. `address(0)`)
     */
    error OwnableInvalidOwner(address owner);

    event OwnershipTransferred(address indexed previousOwner, address indexed newOwner);

    /**
     * @dev Initializes the contract setting the address provided by the deployer as the initial owner.
     */
    constructor(address initialOwner) {
        if (initialOwner == address(0)) {
            revert OwnableInvalidOwner(address(0));
        }
        _transferOwnership(initialOwner);
    }

**Reply by hard1k7724** [reply]
Hey @EngrPips, I think I fixed it. Can you let me know if this is the correct way to do so or not.

pragma solidity ^0.8.18;

import {ERC20Burnable, ERC20} from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";

/*
 * @title DecentralizedStableCoin
 * @author Patrick Collins
 * Collateral: Exogenous (ETH & BTC)
 * Minting: Algorithmic
 * Relative Stability: Pgged to USD
 *
 * This is the contract meant to be governed by the DSCEngine. This contract is just the ERC20 implementation of our stablecoin system.
 *
 */
contract DecentralizedStableCoin is ERC20Burnable, Ownable {
    error DecentralizedStableCoin__MustBeMoreThanZero();
    error DecentralizedStableCoin__BurnAmountExceedsBalance();
    error DecentralizedStableCoin__NotZeroAddress();

    constructor()
        ERC20("DecentralizedStableCoin", "DSC")
        Ownable(address(this))
    {}

I just added the address in the Ownable constructor, There are no errors showing now but please can you tell me if this is the right way or not?

**Reply by EngrPips** [reply, marked as answer]
I assume that would work but maybe setting `msg.sender` would be more appropriate and would enable you to be able to change the owner of the `DecentralizedStableCoin` to `DSCEngine` later on.

**Reply by hard1k7724** [reply]
So  you're saying instead of my contract address(address(this)) I should set it to msg.sender?

**Reply by hard1k7724** [reply]
Also thanks a lot for the help and advice @EngrPips ;))

**Reply by lucaliebenberg** [reply]
Thanks @Juice-07 @EngrPips this has worked for me too.

**Reply by hard1k7724** [reply]
@lucaliebenberg Welcome. I'm glad it did.

**Comment by AKAALBRAMHA** [comment]
Add "abstract" before "contract" in while declaring contract. 
![Screenshot ](https://github.com/Cyfrin/foundry-full-course-f23/assets/103212555/381c4fba-1b0b-46f4-b1c5-7b4a79b407b0)

this worked for me.

**Reply by EngrPips** [reply]
This fixed it, but that is not what you want to achieve. Now, `DecentralizedStableCoin` is abstract and has to be inherited by a child contract. You would actually set the `Owner` property of the `Ownable` contract, and only then will the child contract work properly. Please follow @Juice-07's fix, as that is the proper fix.

**Reply by AKAALBRAMHA** [reply]
Thanks for correcting me.

**Reply by EngrPips** [reply]
You are welcome, friend.

**Reply by AKAALBRAMHA** [reply]
Cool.
